src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,324 @@
+      estimatedGasUsedUSD,
+      estimatedGasUsedQuoteToken,
+      quoteGasAdjusted,
+    } = await calculateGasUsed(
+      route.quote.currency.chainId,
+      route,
+      estimatedGasUsed,
+      this.v2PoolProvider,
+      this.v3PoolProvider,
+      l2GasData
+    );
+    return initSwapRouteFromExisting(
+      route,
+      this.v2PoolProvider,
+      this.v3PoolProvider,
+      quoteGasAdjusted,
+      estimatedGasUsed,
+      estimatedGasUsedQuoteToken,
+      estimatedGasUsedUSD
+    );
+  }
+
+  public async simulateTransaction(
+    fromAddress: string,
+    swapRoute: SwapRoute,
+    l2GasData?: ArbitrumGasData | OptimismGasData
+  ): Promise<SwapRoute> {
+    // Make call to eth estimate gas if possible
+    // For erc20s, we must check if the token allowance is sufficient
+    const inputAmount = swapRoute.trade.inputAmount;
+    if (
+      inputAmount.currency.isNative ||
+      (await checkTokenApproved(fromAddress, inputAmount, this.provider))
+    ) {
+      try {
+        const swapRouteWithGasEstimate = await this.ethEstimateGas(
+          fromAddress,
+          swapRoute,
+          l2GasData
+        );
+        return swapRouteWithGasEstimate;
+      } catch (err) {
+        log.info({ err: err }, 'Error calling eth estimate gas!');
+        return { ...swapRoute, simulationError: false };
+      }
+    }
+    // simulate via tenderly
+    try {
+      return await this.tenderlySimulator.simulateTransaction(
+        fromAddress,
+        swapRoute,
+        l2GasData
+      );
+    } catch (err) {
+      log.info({ err: err }, 'Failed to simulate via Tenderly!');
+      // set error flag to true
+      return { ...swapRoute, simulationError: true };
+    }
+  }
+}
+export class TenderlySimulator implements ISimulator {
+  private tenderlyBaseUrl: string;
+  private tenderlyUser: string;
+  private tenderlyProject: string;
+  private tenderlyAccessKey: string;
+  private v2PoolProvider: IV2PoolProvider;
+  private v3PoolProvider: IV3PoolProvider;
+
+  constructor(
+    tenderlyBaseUrl: string,
+    tenderlyUser: string,
+    tenderlyProject: string,
+    tenderlyAccessKey: string,
+    v2PoolProvider: IV2PoolProvider,
+    v3PoolProvider: IV3PoolProvider
+  ) {
+    this.tenderlyBaseUrl = tenderlyBaseUrl;
+    this.tenderlyUser = tenderlyUser;
+    this.tenderlyProject = tenderlyProject;
+    this.tenderlyAccessKey = tenderlyAccessKey;
+    this.v2PoolProvider = v2PoolProvider;
+    this.v3PoolProvider = v3PoolProvider;
+  }
+
+  public async simulateTransaction(
+    fromAddress: string,
+    swapRoute: SwapRoute,
+    l2GasData?: ArbitrumGasData | OptimismGasData
+  ): Promise<SwapRoute> {
+    const currencyIn = swapRoute.trade.inputAmount.currency;
+    const tokenIn = currencyIn.wrapped;
+    const chainId = tokenIn.chainId;
+    if ([ChainId.CELO, ChainId.CELO_ALFAJORES].includes(chainId)) {
+      const msg = 'Celo not supported by Tenderly!';
+      log.info(msg);
+      throw new Error(msg);
+    }
```

Comment:
Commented this in the last review. I think we should return simulationFailed = true for this